jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,517 @@
+            {
+                _sampleStart = System.nanoTime();
+            }
+
+            @Override
+            public boolean isRateExceeded(long now, boolean addSample)
+            {
+                // Count the request
+                if (addSample)
+                    _sampleCount++;
+
+                long elapsedTime = now - _sampleStart;
+
+                // We calculate the rate if:
+                //    + we didn't add a sample
+                //    + the sample exceeds the rate
+                //    + the sample period has been exceeded
+                if (!addSample || _sampleCount > _maxRequestsPerSecond || (_sampleStart != 0 && elapsedTime > _samplePeriod))
+                {
+                    double elapsedTime1 = (double)(now - _sampleStart);
+                    double count = _sampleCount;
+                    if (elapsedTime1 > 0.0)
+                    {
+                        double currentRate = (count * TimeUnit.SECONDS.toNanos(1L)) / elapsedTime1;
+                        // Adjust alpha based on the ratio of elapsed time to the interval to allow for long and short intervals
+                        double adjustedAlpha = _alpha * (elapsedTime1 / _samplePeriod);
+                        if (adjustedAlpha > 1.0)
+                            adjustedAlpha = 1.0; // Ensure adjustedAlpha does not exceed 1.0
+
+                        _exponentialMovingAverage = (adjustedAlpha * currentRate + (1.0 - adjustedAlpha) * _exponentialMovingAverage);
+                    }
+                    else
+                    {
+                        // assume count as the rate for the sample.
+                        double guessedRate = count * TimeUnit.SECONDS.toNanos(1) / _samplePeriod;
+                        _exponentialMovingAverage = (_alpha * guessedRate + (1.0 - _alpha) * _exponentialMovingAverage);
+                    }
+
+                    // restart the sample
+                    _sampleStart = now;
+                    _sampleCount = 0;
+                }
+
+                // if the rate has been exceeded?
+                return _exponentialMovingAverage > _maxRequestsPerSecond;
+            }
+
+            @Override
+            public boolean isIdle(long now)
+            {
+                return !isRateExceeded(now, false) && _exponentialMovingAverage <= 0.0001;
+            }
+
+            double getCurrentRatePerSecond()
+            {
+                return _exponentialMovingAverage;
+            }
+        }
+    }
+
+    /**
+     * A Handler to reject DOS requests with {@link HttpStatus#ENHANCE_YOUR_CALM_420}.
+     */
+    public static class EnhanceYourCalmRejectHandler implements Request.Handler
+    {
+        @Override
+        public boolean handle(Request request, Response response, Callback callback) throws Exception
+        {
+            Response.writeError(request, response, callback, HttpStatus.ENHANCE_YOUR_CALM_420);
```

Comment:
Maybe this should be 429 Too Many Requests? Because 420 is an unofficial code.

Reply:
awww! you're no fun! but correct!